lib/libspl/atomic.c

```diff
@@ -313,6 +313,19 @@ atomic_swap_ptr(volatile void *target, void *bits)
 	return (__atomic_exchange_n((void **)target, bits, __ATOMIC_SEQ_CST));
 }
 
+#ifndef _LP64
+uint64_t
+atomic_load_64(volatile uint64_t *target)
+{
+	return (__atomic_load_n(target, __ATOMIC_RELAXED));
```

Comment:
Is  `__ATOMIC_RELAXED` really sufficient here or should we be using `__ATOMIC_SEQ_CST`?

https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

Reply:
It depends on expected semantics.  For purposes of just atomically copying the value it is sufficient.  If somebody needs more ordering, then FreeBSD provides additional atomic_load_acq() and atomic_store_rel() primitives.  I don't know why наб put the strictest __ATOMIC_SEQ_CST ordering on all other atomics in atomic.c. I think it is an overkill in attempt to be safe.

Reply:
I'd prefer we agree on semantics and follow it, rather then first overtightening it and then trying to optimize after somebody started to rely on it.  According to https://illumos.org/man/9F/atomic_ops and https://docs.oracle.com/cd/E36784_01/html/E36886/atomic-ops-9f.html  atomics in ZFS originally were not implying any ordering. According to https://www.kernel.org/doc/html/v5.13-rc4/staging/index.html#atomic-types only some of atomic operations imply barriers on Linux by default.  On FreeBSD https://www.freebsd.org/cgi/man.cgi?query=atomic&sektion=9&format=html the required semantics should be specified explicitly and by default it depends on specific platform.

Reply:
To my knowledge nothing is ZFS currently depends on the stricter semantics.  We should be able to adopt the relaxed ordering semantics, if we do so let's just make sure to clearly document what they are since we need to make sure the semantics are the same across platforms.

Reply:
Sounds good to me.  Should we push this PR as-is and then as next PR turn all the atomic.c functions into inlined `__atomic_*(, __ATOMIC_RELAXED)`?

Reply:
Sure, that sounds reasonable to me.  Let's just get a second approval on this, then get it merged.